### PR TITLE
Fix slow chat switching 2: Electric Boogaloo

### PR DESCRIPTION
**Pull request by jrainville**
Redo of https://github.com/status-im/status-desktop/pull/16178

Fixes https://github.com/status-im/status-desktop/issues/16132

In the previous PR, I was improving the slow chat switching by delaying the loading of the member list and the nkeeping it in memory.

After discussing with @micieslak , we agreed that it was making the code harder to understand and also didn't fix the root cause.

Turns out, the root cause is the SortFilterProxyModel and the `preferredDisplayName` property. 

@caybro already has [a PR](https://github.com/status-im/status-desktop/pull/16277) improving it, but I pushed it one step further (sorry)

In my PR, I moved the `preferredDisplayName` property to the Nim side (this is the biggest speed up).

I also made it so that we populate the models in one go instead of one entry at a time (saves just a little bit on start, but it's something)

Finally, I made it so that we don't update the user model if not needed.
Basically, when a community channel is not encrypted is when we put **all** the members of the community in the member list. Those are the channels that are the slowest to switch to.
To help that, I made it so that if we are switching from an unencrypted channel to another unencrypted channel, we just don't update the member list.
This makes the channel switching instantaneous.

The only caveat of that approach is that we are putting a little bit of app logic in the QML. If that's a problem, I can move it to a property in the model.

[fast-switching.webm](https://github.com/user-attachments/assets/73372237-60f6-409e-92ef-55dbc6a78571)

### CODE REVIEW

File: src/app/modules/main/chat_section/chat_content/chat_details.nim
Comment (caybro) [inline src/app/modules/main/chat_section/chat_content/chat_details.nim:326]: There should be a guard against assigning (and reevaluating) the same value

File: ui/app/AppLayouts/Chat/panels/UserListPanel.qml
Comment (caybro) [inline ui/app/AppLayouts/Chat/panels/UserListPanel.qml:108]: Pls leave this part alone :) The StatusMemberListItem does its own composition, depending on both `nickName` and `userName`

**Review by caybro — COMMENTED** [review]
Looks good in general!

File: ui/app/AppLayouts/Chat/panels/UserListPanel.qml
Comment (caybro) [reply ui/app/AppLayouts/Chat/panels/UserListPanel.qml:108]: I started working on the `StatusMemberListItem`, I can take care of this in a followup PR

File: src/app/modules/shared_models/member_model.nim
Comment (caybro) [inline src/app/modules/shared_models/member_model.nim:130]: Actually one more thing: anywhere in this file, whenever you touch any of the 4 properties, you should also emit `dataUpdated(..., ModelRole.PreferredDisplayName.int)`

File: src/app/modules/shared_models/member_model.nim
Comment (alexjba) [inline src/app/modules/shared_models/member_model.nim:122]: nitpick: I'm wondering if there's a reason why it's not a member in the `item` object? 😄

We'll still need to take care of the  `dataChanged` events for this new role and add this logic in all functions updating any of the user names. So IMO adding it as a member in the `item` object would be a nice way to go here.

File: ui/app/AppLayouts/Chat/views/ChatView.qml
Comment (alexjba) [inline ui/app/AppLayouts/Chat/views/ChatView.qml:174]: both `isFullCommunityList` and the `Connections` below have a dependency on the `onChatContentModuleChanged` signal. The implementation assumes that `isFullCommunityList` will always receive the signal first and its value is updated before handling the connection signal.
I'd redo this a little bit to remove any dependency on the signal order.

File: ui/app/AppLayouts/Chat/views/ChatView.qml
Comment (alexjba) [inline ui/app/AppLayouts/Chat/views/ChatView.qml:182]: I'm wondering if this isn't a bit dangerous..What if we delete the previous channel and the model is still cached here? I think it would lead to a crash since there are cases where the `usersModel` won't have a binding anymore on the proper `chatContentModule`, but the model reference assigned below.

**Review by alexjba — COMMENTED** [review]
Nicee! Glad to see this improvement is happening now 👍 

Some doubts below, otherwise looks good to me!

File: src/app/modules/shared_models/member_model.nim
Comment (jrainville) [reply src/app/modules/shared_models/member_model.nim:122]: I prefer adding the `PreferredDisplayName` in `dataChanged` in the model, as it's way simpler than modifying all the modules that use this model. 
Also, this way, the logic for the order which name is more important is only in one place.

File: ui/app/AppLayouts/Chat/views/ChatView.qml
Comment (jrainville) [reply ui/app/AppLayouts/Chat/views/ChatView.qml:182]: I tested it and it didn't crash, but I totally get your worry. I also think it can be dangerous. 
I'm not sure how to make it safe though (apart from just removing this).

I know back in the day when we switched from one UserList for the whole community to having a specific lists per channel, I suggested having one model for all the members of the community that we can show for unpermissioned channels, and then permissioned channels could have their own model, but I know it wasn't done because it was too complex with the current architecture.

I'll try to figure out a way to do it now.
It would free a lot of memory and also fix the worry that it might crash.
I'll open a new issue for it though and work on it in another PR on top

Comment (jrainville) [reply ui/app/AppLayouts/Chat/views/ChatView.qml:182]: Issue opened here https://github.com/status-im/status-desktop/issues/16288

**Review by caybro — APPROVED** [review]
LGTM

File: ui/app/AppLayouts/Chat/views/ChatView.qml
Comment (jrainville) [reply ui/app/AppLayouts/Chat/views/ChatView.qml:182]: PR opened here: https://github.com/status-im/status-desktop/pull/16301

We can wait until that new PR is merged before merging this one, that way we don't have to merge the possibly dangerous code.

Comment (jrainville) [reply ui/app/AppLayouts/Chat/views/ChatView.qml]: Alright, I removed the dangerous code. I was tired of rebasing lal my PRs on this. I want to merge it hahaha

File: ui/app/AppLayouts/Chat/views/ChatView.qml
Comment (micieslak) [inline ui/app/AppLayouts/Chat/views/ChatView.qml:175]: is it needed?

File: src/app/modules/shared_models/member_model.nim
Comment (micieslak) [inline src/app/modules/shared_models/member_model.nim:191]: As we have `ModelCount` attached property, there is no need to expose `count` from nim. Ideally we should just remove it from the model. Not necessary in this PR, just pointing for the future.

File: src/app/modules/shared_models/member_model.nim
Comment (micieslak) [inline src/app/modules/shared_models/member_model.nim:181]: please return early when `items` is empty. Otherwise we can end up with malformed notifications (e.g. `start = 0` and `end = -1`)

File: src/app/modules/shared_models/member_model.nim
Comment (micieslak) [inline src/app/modules/shared_models/member_model.nim:322]: I think it would be beneficial to check if any name used for computing preferred display name changed and only then included that role. It's quite important as that role is used for sorting and notifying change on it will trigger sorting.

**Review by micieslak — COMMENTED** [review]
Nice work! I'm happy to see such simplifications and performance boost in the same time!

I left some minor comments/suggestions only.

**Review by micieslak — APPROVED** [review]
Nice!

**Pull request merged**
